### PR TITLE
Implement method to return bucket type from server

### PR DESCRIPTION
**Pull request by Tulsishah**
### Description
Implemented FetchAndSetBucketType() method to get bucket type from `GetStorageLayout`(e.g. Hierarchical or NonHierarchical) and set it in bucketType.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Added(Create mock control client)
3. Integration tests - Added

### CODE REVIEW

File: internal/storage/storage_handle_test.go [hunk 1]
```diff
@@ -298,3 +303,10 @@ func (testSuite *StorageHandleTest) TestCreateClientOptionForGRPCClient() {
 	assert.Nil(testSuite.T(), err)
 	assert.NotNil(testSuite.T(), clientOption)
 }
+
+// EnableHNS flag is false so conrtolClient will be nil.
```
Comment (raj-prince) [inline internal/storage/storage_handle_test.go:307]: nit: controlClient.

File: internal/gcsx/bucket_manager.go [hunk 2]
```diff
@@ -222,6 +222,9 @@ func (bm *bucketManager) SetUpBucket(
 		bm.config.TmpObjectPrefix,
 		b)
 
+	// Fetch and set bucket type
+	b.FetchAndSetBucketType()
```
Comment (raj-prince) [inline internal/gcsx/bucket_manager.go:226]: This will require additional permission to the user while mounting, right?

Do we need to update any documentation or playbook?

**Review by raj-prince — APPROVED** [review]

File: internal/storage/bucket_handle.go [hunk 3]
```diff
@@ -430,6 +435,35 @@ func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjec
...
+	}
+
+	var callOptions []gax.CallOption
+	stoargeLayout, err := b.controlClient.GetStorageLayout(context.Background(), &controlpb.GetStorageLayoutRequest{
+		Name:      "projects/_/buckets/" + b.bucketName + "/storageLayout",
+		Prefix:    "",
+		RequestId: "",
+	}, callOptions...)
+
+	// In case bucket does not exist, set type unknown instead of panic.
```
Comment (raj-prince) [inline internal/storage/bucket_handle.go:452]: What happens if error comes because of some permission issue?

File: internal/storage/bucket_handle.go [hunk 4]
```diff
@@ -430,6 +435,35 @@ func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjec
...
+	if err != nil {
+		b.bucketType = gcs.Unknown
+		log.Printf("GetStorageLayout: %v", err)
+		return
+	}
+
+	if stoargeLayout.HierarchicalNamespace != nil && stoargeLayout.GetHierarchicalNamespace().Enabled {
```
Comment (raj-prince) [inline internal/storage/bucket_handle.go:459]: It's better to get HierarchicalNamespace via method instead of accessing directly.

```
hierarchicalNamespace := stoargeLayout.GetHierarchicalNamespace()
if hierarchicalNamespace != nil && hierarchicalNamespace.Enabled {
...
}
```

File: internal/storage/gcs/bucket.go [hunk 5]
```diff
@@ -44,6 +45,9 @@ type Bucket interface {
 	// Return Type of bucket e.g. Hierarchical or NonHierarchical
 	BucketType() BucketType
 
+	// Fetch bucket type from GetStorageLayout and set it. e.g. Hierarchical or NonHierarchical
```
Comment (raj-prince) [inline internal/storage/gcs/bucket.go:48]: Is it necessary to add a new method in the bucket interface?

Can't we set at the first BucketType() method call, or during initialization of bucket handle?

**Review by raj-prince — COMMENTED** [review]
Requires discussion.

File: internal/storage/bucket_handle.go (hunk 3 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle.go:452]: `storage.objects.list` permission is require for this API call. This permission is included in `Storage Object Viewer ` role. Which is I think minimum role require to mount the bucket.

If user does not have `Storage Object Viewer `  role then anyway it will fail on very first List call.

File: internal/storage/gcs/bucket.go (hunk 5 above)
Comment (Tulsishah) [reply internal/storage/gcs/bucket.go:48]: All the clients and control client method are in bucket_handle and get called by  bucket_handle interface only. So to keep consistency we decided that.

File: internal/gcsx/bucket_manager.go [hunk 6]
```diff
@@ -222,6 +222,9 @@ func (bm *bucketManager) SetUpBucket(
 		bm.config.TmpObjectPrefix,
 		b)
 
+	// Fetch and set bucket type
```
Comment (ankitaluthra1) [inline internal/gcsx/bucket_manager.go:225]: update to "Fetch bucket type from storage layout api and set bucket type"

File: internal/storage/bucket_handle.go [hunk 7]
```diff
@@ -430,6 +435,36 @@ func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjec
...
+	stoargeLayout, err := b.controlClient.GetStorageLayout(context.Background(), &controlpb.GetStorageLayoutRequest{
+		Name:      "projects/_/buckets/" + b.bucketName + "/storageLayout",
+		Prefix:    "",
+		RequestId: "",
+	}, callOptions...)
+
+	// In case bucket does not exist, set type unknown instead of panic.
+	if err != nil {
+		b.bucketType = gcs.Unknown
+		log.Printf("GetStorageLayout: %v", err)
```
Comment (ankitaluthra1) [inline internal/storage/bucket_handle.go:455]: log should be of error type and update statement to "Error returned from GetStorageLayout: %v", err"

File: internal/storage/bucket_handle.go [hunk 8]
```diff
@@ -430,6 +435,36 @@ func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjec
...
+		return
+	}
+
+	hierarchicalNamespace := stoargeLayout.GetHierarchicalNamespace()
+	if hierarchicalNamespace != nil && hierarchicalNamespace.Enabled {
+		b.bucketType = gcs.Hierarchical
+		return
+	}
+
+	b.bucketType = gcs.Unknown
```
Comment (ankitaluthra1) [inline internal/storage/bucket_handle.go:465]: in case hierarchicalNamespace is nil or hierarchicalNamespace is not Enabled, bucket type should be NonHierarchical not unknown

File: internal/storage/bucket_handle_test.go [hunk 9]
```diff
@@ -1187,3 +1192,26 @@ func (testSuite *BucketHandleTest) TestComposeObjectMethodWithOneSrcObjectIsDstO
...
+func (testSuite *BucketHandleTest) TestFetchBucketTypeWithError() {
+	var x *controlpb.StorageLayout
+	// Test when the client returns an error.
+	testSuite.mockClient.On("GetStorageLayout", mock.Anything, mock.Anything, mock.Anything).
+		Return(x, errors.New("mocked error"))
+
+	bucketType := mockFetchAndSetBucketType(testSuite.mockClient)
+
+	assert.Equal(testSuite.T(), gcs.Unknown, bucketType, "Expected Unknown when there's an error")
+}
```
Comment (ankitaluthra1) [inline internal/storage/bucket_handle_test.go:1217]: These two tests are testing mockFetchAndSetBucketType, actual FetchAndSetBucketType method which will be used in production code is not being tested here. To validate FetchAndSetBucketType try to mock only client and create new bucket hadnle with embedded mock client and in the test actual FetchAndSetBucketType method should be called. 

Also, two more test cases needs to be added here, one when control client is nil and second when storagelayout does not give any error but hierarchial block is not present in success response.

File: internal/storage/bucket_handle.go (hunk 3 above)
Comment (ankitaluthra1) [reply internal/storage/bucket_handle.go:452]: @raj-prince in that scenario also, bucket type should be set as unknown, and it will be treated as flat bucket

Comment (ankitaluthra1) [reply internal/storage/bucket_handle.go:452]: @Tulsishah "If user does not have Storage Object Viewer  role then anyway it will fail on very first List call" --> as I can see in this PR, set type getting called before first list call inside setup bucket ? are we doing list call before setup bucket as well ?

File: internal/storage/gcs/bucket.go (hunk 5 above)
Comment (ankitaluthra1) [reply internal/storage/gcs/bucket.go:48]: can we call FetchAndSetBucketType [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/f86b35dd7d21b90d0b41fda56e736a310304e881/internal/gcsx/bucket_manager.go#L167), before b is converted to corresponding bucket, that way only bucket handle needs to have method, thoughts ? @Tulsishah @raj-prince

File: internal/storage/bucket_handle_test.go (hunk 9 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle_test.go:1239]: Done please check.

File: internal/storage/gcs/bucket.go (hunk 5 above)
Comment (Tulsishah) [reply internal/storage/gcs/bucket.go:48]: No we can't do that because b type is gcs.Bucket. So it have to part of bucket interface to call it from setUpBucket.

Comment (ankitaluthra1) [reply internal/storage/gcs/bucket.go:48]: got it, lets keep it in all bucket types then, regarding two alternatives:
set at first buckettype --> this will slow down the first corresponding operation which will trigger bucket type, which is not expected
initialization of bucket handle--> bucket handle is being initialised in storage handle, storage handle is not the correct domain to call storage layout api, setup bucket method should only know how to fetch type and set for the bucket, this behaviour should be abstracted from storage handl, keeping it in storage handle breaks abstraction principle.

**Review by ankitaluthra1 — DISMISSED** [review]
lgtm

File: internal/storage/gcs/bucket.go (hunk 5 above)
Comment (Tulsishah) [reply internal/storage/gcs/bucket.go]: Had offline discussion to merge method in BucketType() only.

File: internal/storage/bucket_handle.go [hunk 10]
```diff
@@ -27,24 +27,55 @@ import (
...
+	bucketName    string
+	bucketType    gcs.BucketType
+	controlClient StorageControlClientInterface
 }
 
 func (bh *bucketHandle) Name() string {
 	return bh.bucketName
 }
 
 func (bh *bucketHandle) BucketType() gcs.BucketType {
```
Comment (raj-prince) [inline internal/storage/bucket_handle.go:53]: Can we please write a comment here, similar to below:
```During the first BucketType() call, type is inferred by making a GCS call and cached in a local variable. So, that the first call will be a little slower, and further call will be faster since no GCS call.

Also, this is thread safe, but parallelly calling by mutiple threads the first time might result in the more than one GCS call. So, it's good to call it while mounting first to avoid this case.
```




(

File: internal/storage/fake/bucket.go [hunk 11]
```diff
@@ -440,6 +440,10 @@ func (b *bucket) BucketType() gcs.BucketType {
 	return b.bucketType
 }
 
+func (b *bucket) FetchAndSetBucketType() {
```
Comment (raj-prince) [inline internal/storage/fake/bucket.go:443]: We need to revert this change.

File: internal/storage/control_client_wrapper.go [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"context"
+
+	control "cloud.google.com/go/storage/control/apiv2"
+	"cloud.google.com/go/storage/control/apiv2/controlpb"
+	"github.com/googleapis/gax-go/v2"
+)
+
+// Define the interface
```
Comment (raj-prince) [inline internal/storage/control_client_wrapper.go:25]: I don't see this comment has a lot of value here.

Can we be specific about this interface, if not clearly visible by the name?

File: internal/storage/control_client_wrapper.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+type StorageControlClient interface {
+	GetStorageLayout(ctx context.Context,
+		req *controlpb.GetStorageLayoutRequest,
+		opts ...gax.CallOption) (*controlpb.StorageLayout, error)
+}
+
+type StorageControlClientWrapper struct {
```
Comment (raj-prince) [inline internal/storage/control_client_wrapper.go:32]: I don't see any use-case of this wrapper class. If not used, can we revert this?

File: internal/storage/control_client_wrapper.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"context"
+
+	control "cloud.google.com/go/storage/control/apiv2"
+	"cloud.google.com/go/storage/control/apiv2/controlpb"
+	"github.com/googleapis/gax-go/v2"
+)
+
+// Define the interface
+type StorageControlClient interface {
```
Comment (raj-prince) [inline internal/storage/control_client_wrapper.go:26]: Also, could you please check with go-sdk team? Ideally, it would be better to keep this interface go-sdk side.

File: internal/storage/mock_control_client.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"context"
+
+	"cloud.google.com/go/storage/control/apiv2/controlpb"
+	"github.com/googleapis/gax-go/v2"
+	"github.com/stretchr/testify/mock"
+)
+
+// MockStorageControlClient creates a mock version of the StorageControlClient.
+type MockStorageControlClient struct {
```
Comment (raj-prince) [inline internal/storage/mock_control_client.go:26]: should we keep `StorageControlClient` inside this struct? Just to make sure, this mock is implementing all the methods of `StorageControlClient`.

File: internal/storage/control_client_wrapper.go (hunk 14 above)
Comment (Tulsishah) [reply internal/storage/control_client_wrapper.go:25]: Sure will check with them

File: internal/storage/control_client_wrapper.go (hunk 13 above)
Comment (Tulsishah) [reply internal/storage/control_client_wrapper.go]: Done

**Review by raj-prince — COMMENTED** [review]
Went through with the interface level change. Added a couple of comments.

File: internal/storage/gcs/bucket.go [hunk 16]
```diff
@@ -26,8 +26,10 @@ type BucketType int
 
 // BucketType enum values.
 const (
-	NonHierarchical BucketType = iota
+	Nil BucketType = iota
```
Comment (ankitaluthra1) [inline internal/storage/gcs/bucket.go:29]: please add a comment here Default value Nil represents that bucket type is not set

**Review by ankitaluthra1 — DISMISSED** [review]
lgtm

**Review by raj-prince — APPROVED** [review]
Relying on Ankita's review for unit test changes.
